staging/kos/pkg/controllers/subnet/validator.go

```diff
@@ -228,33 +227,33 @@ func (v *Validator) processDeletedSubnet(s k8stypes.NamespacedName) {
 func (v *Validator) processExistingSubnet(s *netv1a1.Subnet) error {
 	ss, parsingErrs := subnet.NewSummary(s)
 	if len(parsingErrs) > 0 {
-		return parsingErrs
+		glog.Errorf("Subnet %s/%s is malformed: %s. This should never happen.", s.Namespace, s.Name, parsingErrs.Error())
+		return nil
 	}
 
 	if v.subnetIsStale(ss.NamespacedName, s.ResourceVersion) {
-		return nil
+		return fmt.Errorf("current version (%s) is stale", s.ResourceVersion)
```

Comment:
Returning nil was a bug because it assumed a new reference to the attachment would be enqueued upon receiving the fresh version of the subnet. But the fresh version of the subnet is delivered through an update notification which does not perform an enqueue, because enqueues are conditional to a change in UID (or  VNI and CIDR before this PR).

Reply:
(A) requeuing in this case is effectively polling for a notification.  Normally the notification should come quickly --- perhaps even more quickly than the retry; in bad weather the notification might not arrive for a while.  While not incorrect, I suspect polling for notification is not the best approach.

(B) Since a SubnetSpec is entirely immutable, the Subnet can need new processing only if the UID has changed, right?

Reply:
I'll give a unique answer for (A) and (B).

A subnet X needs re-processing also if a rival has been deleted.
This is achieved by recording X in a rival Y's conflicts cache. Upon Y's deletion X is enqueued and reprocessed. But if X was in Y's conflicts cache there's a conflict and its `status.errors` field has been updated to reflect that. If Y is deleted and X is re-processed before X's updated version appears in the Informer's cache, re-processing of X will fail because the new status update will fail, the cause being a resource version mismatch. This is what happens if we remove entirely the check. It is not incorrect: the error from the server will cause a delayed re-processing of the subnet and hopefully by then the Informer's cache records the up to date version of X.

Having that check (and returning an error if the subnet is stale) achieves the exact same result without the call to the API server, which is good for performance (the idea I got by reading PRs and issues in K8s and docs is that K8s people are rather protective with the API servers --- if a call can be easily avoided it should be avoided). OTOH, conflicts should be extremely rare and are always caused by a user's mistake.

You're right that retrying is polling for a notification. Not retrying though is incorrect because the notification will not enqueue anything because UID did not change: a needed processing is never performed. A possible path out of this could be changing the condition in the update notification handler which decides when there is an enqueue, but I don't like the idea and that is still polling for a notification.

Reply:
Right.

The suppression of work on stale objects is correct, because any work on a stale object is guaranteed to fail.

In the handler for an update notification, we have been focused on the question of whether the update indicates a change in object content that matters.  But an update also indicates a change in ResourceVersion, and that alone can matter.  A correct test in the update notification handler would be whether either (a) there is a change in UID or (b) the object is in the stale cache with a ResourceVersion different from the new ResoureVersion.

Reply:
The problem with your solution (the liberal one) is that a conflicting subnet X keeps being validated indefinitely. Its status is updated to record the conflicts, the stale RV is written in the cache. When the update notification is received the subnet namespaced name is enqueued and the subnet is processed again, and this keeps happening indefinitely because there will be a new status update. A way out is the conservative solution you outlined, but as you said it is complex (much more than the current solution --- returning an error and retrying). Also, such solution is not conceptually better than polling for a notification: you're waiting for a specific notification. You're stating that specific events happened and you're waiting for a notification which lets you process them. Another way out is making the status update conditional to whether the update would actually change the status, and if that's not the case returning nil and stopping processing. This way an already updated conflicting subnet is processed only once more than necessary. But the status update is the last step of processing: all the work has already been done (including the expensive list to the API server). I'd rather have work suppression suppress work ASAP.

Reply:
I've been thinking about how to implement the conservative solution. Currently the stale RVs cache is a map from namespaced name to stale RV. Replace the map value with a struct storing the stale RV and a bool saying whether the namespaced name has been evicted from a conflicts cache. The bool is set to true when the subnet is evicted from a conflicts cache. The update notification handler enqueues if the UIDs differ or if the subnet is not stale and the eviction flag is true.

I'm not against leaving things the way they are (returning an error in case of stale subnet AKA polling for a notification). But if you are convinced that we should do something different, I don't think the more conservative solution you outlined adds that much complexity.

Reply:
You touched on another solution that is simple and principled: request an update only if there is actually a change being requested.  Yes, that lets in one round of unnecessary work.  But that should be rare (because conflicts are rare), so I am inclined to stop there.  Leave the update notification handler to simply test whether (a) the UID has changed or (b) there is a doomed entry in the stale RV cache.

Reply:
Done in dc54563.